specifications/cloud_assembly.md

```diff
@@ -0,0 +1,398 @@
+of the archive are reserved for special use:
+* `manifest.json` **MUST** be present and contains the [manifest document](#manifest-document) for the *Cloud Assembly*.
+* `signature.asc`, when present, **MUST** contain the [digital signature](#digital-signature) of the *Cloud Assembly*.
+
+### Manifest Document
+The `manifest.json` file is the entry point of the *Cloud Assembly*. It **MUST** be a valid [JSON] document composed of
+a single `object` that conforms to the following schema:
+
+Key           |Type                 |Required|Description
+--------------|---------------------|:------:|-----------
+`schema`      |`string`             |Required|The schema for the document. **MUST** be `cloud-assembly/1.0`.
+`drops`       |`Map<ID,Drop>`       |Required|A mapping of [*Logical ID*](#logical-id) to [Drop](#drop).
+`missing`     |`Map<string,Missing>`|        |A mapping of context keys to [missing information](#missing).
+
+The [JSON] specification allows for keys to be specified multiple times in a given `object`. However, *Cloud Assembly*
+consumers **MAY** assume keys are unique, and [*Cloud Assemblers*](#cloud-assemblers) **SHOULD** avoid generating
+duplicate keys. If duplicate keys are present, the latest specified value **SHOULD** be preferred.
+
+### Logical ID
+*Logical IDs* are `string`s that uniquely identify [Drop](#drop)s in the context of a *Cloud Assembly*.
+* A *Logical ID* **MUST NOT** be empty.
+* A *Logical ID* **SHOULD NOT** exceed `256` characters.
+* A *Logical ID* **MUST** be composed of only the following ASCII printable characters:
+  + Upper-case letters: `A` (`0x41`) through `Z` (`0x5A`)
+  + Lower-case letters: `a` (`0x61`) through `z` (`0x7A`)
+  + Numeric characters: `0` (`0x30`) through `9` (`0x39`)
+  + Plus: `+` (`0x2B`)
+  + Minus: `-` (`0x2D`)
+  + Forward-slash: `/` (`0x2F`)
+  + Underscore: `_` (`0x5F`)
+* A *Logical ID* **MUST NOT** contain the `.` (`0x2E`) character as it is used in the string substitution pattern for
+  cross-drop references to separate the *Logical ID* from the *attribute* name.
+
+In other words, *Logical IDs* are expected to match the following regular expression:
+```js
+/^[A-Za-z0-9+\/_-]{1,256}$/
+```
+
+### Drop
+Clouds are made of Drops. Thet are the building blocks of *Cloud Assemblies*. They model a part of the
+*cloud application* that can be deployed independently, provided it's dependencies are fulfilled. Drops are specified
+using [JSON] objects that **MUST** conform to the following schema:
+
+Key          |Type             |Required|Description
+-------------|-----------------|:------:|-----------
+`type`       |`string`         |Required|The [*Drop Type*](#drop-type) specifier of this Drop.
+`environment`|`string`         |required|The target [environment](#environment) in which Drop is deployed.
+`metadata`   |`Map<string,any>`|        |Arbitrary key-value pairs associated with this Drop.
```

Comment:
Would be worthwhile defining some standard metadata.

For example, **warnings** and **errors** seem very useful to standardize, as well as the behavior associated with them (errors would block deployment, warnings would cause a prompt, etc.)

Reply:
I agree. Although IMO, "errors" should prevent the cloud assembly from even being created...

Reply:
A single Cloud Assembly can contain multiple drops, one of which contains synthesis errors. That need not prohibit deployment of the other drops.

Or something similar. But we need a way to express this, see https://github.com/awslabs/aws-cdk/issues/1035

Reply:
To me this feels akin to your linker produces a binary that includes some static libraries that had completion errors, and you're asking for the executable to be produced, so you can... maybe attempt to use the other features? I would argue that drops that contain errors should then... be omitted from the assembly entirely.